File: 2023-2024/sourcecode/mod3/lab9/RoomBasicsJava/app/src/main/java/com/example/roombasicsjava/viewmodel/UserViewModel.java
package com.example.roombasicsjava.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.roombasicsjava.data.User;
import com.example.roombasicsjava.data.UserDao;
import com.example.roombasicsjava.data.UserDatabase;
import com.example.roombasicsjava.repository.UserRepository;
import com.example.roombasicsjava.util.AppExecutors;

import java.util.List;

public class UserViewModel extends AndroidViewModel {
    private final UserRepository repository;
    private LiveData<List<User>> users;

    public UserViewModel(@NonNull Application application) {
        super(application);
        UserDatabase db = UserDatabase.getDatabase(application);
        UserDao dao = db.userDao();
        repository = new UserRepository(dao);

        fetchUsers();
    }

    private void fetchUsers() {
        users = repository.getUsers();
    }

    public LiveData<List<User>> getUsers() {
        return users;
    }

    public void clearUsers() {
        AppExecutors.getInstance().diskIO().execute(repository::clear);
    }

    public void addUser(User user) {
        AppExecutors.getInstance().diskIO().execute(() -> repository.add(user));
    }
}

File: 2023-2024/sourcecode/mod2/lab9/RoomBasicsKotlin/app/src/main/java/com/example/roombasicskotlin/viewmodel/UserViewModel.kt
package com.example.roombasicskotlin.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.roombasicskotlin.data.User
import com.example.roombasicskotlin.data.UserDatabase
import com.example.roombasicskotlin.repository.UserRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class UserViewModel(application: Application) : AndroidViewModel(application) {

    private val repository: UserRepository
    private val _usersState = MutableStateFlow<List<User>>(emptyList())
    val usersState: StateFlow<List<User>>
        get() = _usersState

    init {
        val db = UserDatabase.getDatabase(application)
        val dao = db.userDao()
        repository = UserRepository(dao)

        fetchUsers()
    }

    private fun fetchUsers() {
        viewModelScope.launch {
            repository.getUsers().collect { users ->
                _usersState.value = users
            }
        }
    }

    fun clearUsers() {
        viewModelScope.launch {
            repository.clear()
        }
    }

    fun addUser(user: User) {
        viewModelScope.launch {
            repository.add(user)
        }
    }
}